datafusion/src/physical_plan/metrics.rs

```diff
@@ -0,0 +1,461 @@
+    }
+
+    /// Add the standard name for output rows
+
+    /// Get the current value
+    pub fn value(&self) -> usize {
+        self.value.load(Ordering::Relaxed)
+    }
+
+    /// get the kind of this metric
+    pub fn kind(&self) -> &MetricKind {
+        &self.kind
+    }
+
+    /// Add `n` to the metric's value
+    pub fn add(&self, n: usize) {
+        // relaxed ordering for operations on `value` poses no issues
+        // we're purely using atomic ops with no associated memory ops
+        self.value.fetch_add(n, Ordering::Relaxed);
+    }
+
+    /// Set the metric's value to `n`
+    pub fn set(&self, n: usize) {
+        self.value.store(n, Ordering::Relaxed);
+    }
+
+    /// What labels are present for this metric?
+    fn labels(&self) -> &[Label] {
+        &self.labels
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+/// How should the value of the metric be interpreted?
+pub enum MetricKind {
+    /// Number of output rows produced
+    OutputRows,
+    /// CPU time
+    CPUTime,
+    // TODO timestamp, etc
+    // https://github.com/apache/arrow-datafusion/issues/866
+    /// Arbitarary user defined type
+    Custom(Arc<str>),
+}
+
+impl Display for MetricKind {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.name())
+    }
+}
+
+impl MetricKind {
+    /// return a user displayable name of this kind of metric
+    pub fn name(&self) -> &str {
+        match self {
+            MetricKind::OutputRows => "outputRows",
+            MetricKind::Custom(name) => name,
+            MetricKind::CPUTime => "cpuTime",
+        }
+    }
+}
+
+/// A set of SQLMetrics for a particular operator
+#[derive(Default, Debug, Clone)]
+pub struct MetricsSet {
+    metrics: Vec<Arc<SQLMetric>>,
+}
+
+impl MetricsSet {
+    /// Create a new container of metrics
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Add the specified metric
+    pub fn push(&mut self, metric: Arc<SQLMetric>) {
+        self.metrics.push(metric)
+    }
+
+    /// Add all [`SQLMetric`]s in this set to the specified array.
+    fn extend_other(&mut self, metrics: &mut Vec<Arc<SQLMetric>>) {
+        metrics.extend(self.metrics.iter().cloned())
+    }
+
+    /// convenience: return the number of rows produced, aggregated
+    /// across partitions or None if no metric is present
+    pub fn output_rows(&self) -> Option<usize> {
+        self.sum(|metric| matches!(metric.kind(), MetricKind::OutputRows))
+    }
+
+    /// convenience: return the amount of CPU time spent, aggregated
+    /// across partitions or None if no metric is present
+    pub fn cpu_time(&self) -> Option<usize> {
+        self.sum(|metric| matches!(metric.kind(), MetricKind::CPUTime))
+    }
+
+    /// Sums the values for metrics for which `f(metric)` returns
+    /// true, and returns the value. Returns None if no metrics match
+    /// the predicate.
+    pub fn sum<F>(&self, mut f: F) -> Option<usize>
```

Comment:
Here are some of the aggregation primitives (sum and group by partition). I feel this API may grow as we understand the usecases more